File: sort/quickSort.js
'use strict';
const BaseArray = require('./ArrayList.class')
class ArrayList extends BaseArray { 

  quickSort() {
    return this.quick(this.array,)
  }
  quick (array, left, right){

    var index; //{1}
  
    if (array.length > 1) { //{2}
  
      index = this.partition(array, left, right); //{3}
  
      if (left < index - 1) {                //{4}
        quick(array, left, index - 1);     //{5}
      }
  
      if (index < right) {  //{6}
        quick(array, index, right);        //{7}
      }
    }
  }

  partition (array, left, right) {
　
    var pivot = this.array[Math.floor((right + left) / 2)], //{8}
      i = left,                                      //{9}
      j = right;                                     //{10}
      while (i <= j) {                //{11}
      while (array[i] < pivot) {  //{12}
        i++;
      }
      while (array[j] > pivot) {  //{13}
        j--;
      }
      if (i <= j) { //{14}
        this.swap(i, j); //{15}
        i++;
        j--;
      }
    }
    return i; //{16}
  }
}



function ArrayList()
{
  let array = [];
  this.insert = (item) => {
    array.push(item);
  }
  this.toString = () => {
    return array.join();
  }
  this.quickSort =  (array, left, right) => {

    var index; //{1}
  
    if (array.length > 1) { //{2}
  
      index = partition(array, left, right); //{3}
  
      if (left < index - 1) {                //{4}
        quickSort(array, left, index - 1);     //{5}
      }
  
      if (index < right) {  //{6}
        quickSort(array, index, right);        //{7}
      }
    }
  }
}

//TestCode
let arr = new ArrayList();
arr.insert(12);
arr.insert(2);
arr.insert(7);
arr.insert(71);
arr.insert(87);
console.log(arr.toString());
arr.insertionSort();
console.log(arr.toString());

File: go-leetcodecollection/sort/quick_sort.go
package sort

func quickSort(nums []int) {
	sort(nums, 0, len(nums)-1)
}

func sort(nums []int, left int, right int) {
	if right <= left {
		return
	}

	p := partition(nums, left, right) // 获取分区点
	sort(nums, left, p-1)
	sort(nums, p+1, right)

}

func partition(nums []int, left, right int) int {
	pivot := nums[right]
	i := left
	for j := i; j < right; j++ {
		if nums[j] < pivot {
			nums[i], nums[j] = nums[j], nums[i]
			i++
		}
	}
	// 最后将 pivot 与 i 下标对应数据值互换
	// 这样一来，pivot 就位于当前数据序列中间，i 也就是 pivot 值对应的下标
	nums[i], nums[right] = pivot, nums[i]

	// 返回 i 作为 pivot 分区位置
	return i
}
